### PR TITLE
Introduce websocket to fix the bug that project config not sync with the server

### PR DESCRIPTION
**Pull request by NingW101**
please test with https://github.com/intel/webinizer-webclient/pull/30

### CODE REVIEW

**Comment by NingW101** [comment]
#### Some explanation and description: 

The key considerations to introduce WS to solve the bugs where the client configuration information is out of sync with the server

1. create websocket connections only when necessary, remove connections when they are unnecessary.
2. accurately send the WS messages to the clients that need updated information, try not send the messages to the clients that not need to update info.

As for 1st point, in front-end we create try to create websocket connection in `onMounted` hook in the `Build` page and `Configure` page, and close the connection in `onUnmounted` hook. Now we do not differentiate between projects that have dependencies or whether they are added as dependencies in other projects, because it hard to know if one project is added as dependency. 

As for 2nd point, in back-end we maintain a map to store the projects and their connected websocket connections. For the scenarios that one dependency is triggered building process from the main project, we will send the the message through this dependency's connected ws connections to the clients and make them update the config. For situations where the "Build Target" and "Package Information" in dependencies change without the main project knowing about it, we broadcast to all 
 connections since it's hard to know who a project is used as a dependency. We also define the message type in the parameter `wsMsgType` inside of the WS message content, and the clients(configure page) will check the type then handle it properly accordingly.

File: src/project.ts [hunk 1]
```diff
@@ -317,8 +318,26 @@ export class Project implements IProject {
           depBuildStatus !== "building" &&
           depBuildStatus !== "building_with_recipes"
         ) {
-          log.info(`... build dependency: ${k}@${depRoot}`, dumpLog);
+          log.info(`... build dependency: ${k}@${depRoot}`, dumpLog); // there is a scenario that the dependency cannot be
```
Comment (lingyuncai) [inline src/project.ts:321]:
```suggestion
          log.info(`... build dependency: ${k}@${depRoot}`, dumpLog);
          // there is a scenario that the dependency cannot be
```

File: src/ws/websocket.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        log.warn("the socket request is from the project - ", root);
+        this.webSocketServer?.handleUpgrade(req, socket, head, (socket) => {
+          this.webSocketServer?.emit("connection", socket, root);
+        });
+      });
+    }
+  }
+
+  onConnectionHandler = (webSocketConnection: ws, root: string) => {
+    // this.addConnection(webSocketConnection);
```
Comment (lingyuncai) [inline src/ws/websocket.ts:63]:
```suggestion
```

**Review by lingyuncai — APPROVED** [review]
LGTM, thanks Ning~

**Pull request merged**